hack/make.sh

```diff
@@ -83,53 +82,58 @@ add_buildtag() {
 	[[ " $DOCKER_BUILDTAGS" == *" $1_"* ]] || DOCKER_BUILDTAGS+=" $1_$2"
 }
 
-if ${PKG_CONFIG} 'libsystemd' 2> /dev/null; then
-	DOCKER_BUILDTAGS+=" journald"
-fi
-
-# test whether "libdevmapper.h" is new enough to support deferred remove
-# functionality. We favour libdm_dlsym_deferred_remove over
-# libdm_no_deferred_remove in dynamic cases because the binary could be shipped
-# with a newer libdevmapper than the one it was built with.
-if
-	command -v gcc &> /dev/null \
-		&& ! (echo -e '#include <libdevmapper.h>\nint main() { dm_task_deferred_remove(NULL); }' | gcc -xc - -o /dev/null $(pkg-config --libs devmapper) &> /dev/null) \
-		;
-then
-	add_buildtag libdm dlsym_deferred_remove
+if [ -z "$CGO_ENABLED" ]; then
+	case "$(go env GOOS)/$(go env GOARCH)" in
+		darwin/* | windows/amd64 | linux/amd64 | linux/arm64 | linux/arm | linux/s390x | linux/ppc64le | linux/riscv*)
+			export CGO_ENABLED=1
+			;;
+		*)
+			export CGO_ENABLED=0
+			;;
+	esac
 fi
 
-# Use these flags when compiling the tests and final binary
-
-IAMSTATIC='true'
-if [ -z "$DOCKER_DEBUG" ]; then
-	LDFLAGS='-w'
+if [ "$CGO_ENABLED" = "1" ] && [ "$DOCKER_STATIC" = "1" ] && [ "$(go env GOOS)" = "linux" ]; then
+	DOCKER_LDFLAGS+=" -extldflags -static"
 fi
 
-LDFLAGS_STATIC=''
-EXTLDFLAGS_STATIC='-static'
-# ORIG_BUILDFLAGS is necessary for the cross target which cannot always build
-# with options like -race.
-ORIG_BUILDFLAGS=(-tags "netgo osusergo static_build $DOCKER_BUILDTAGS" -installsuffix netgo)
-# see https://github.com/golang/go/issues/9369#issuecomment-69864440 for why -installsuffix is necessary here
-
-BUILDFLAGS=(${BUILDFLAGS} "${ORIG_BUILDFLAGS[@]}")
-
-LDFLAGS_STATIC_DOCKER="
-	$LDFLAGS_STATIC
-	-extldflags \"$EXTLDFLAGS_STATIC\"
-"
-
 if [ "$(uname -s)" = 'FreeBSD' ]; then
 	# Tell cgo the compiler is Clang, not GCC
 	# https://code.google.com/p/go/source/browse/src/cmd/cgo/gcc.go?spec=svne77e74371f2340ee08622ce602e9f7b15f29d8d3&r=e6794866ebeba2bf8818b9261b54e2eef1c9e588#752
 	export CC=clang
 
 	# "-extld clang" is a workaround for
 	# https://code.google.com/p/go/issues/detail?id=6845
-	LDFLAGS="$LDFLAGS -extld clang"
+	DOCKER_LDFLAGS+=" -extld clang"
+fi
+
+if [ "$CGO_ENABLED" = "1" ] && [ "$DOCKER_STATIC" = "1" ]; then
+	DOCKER_BUILDTAGS+=" netgo osusergo static_build"
 fi
 
+if [ "$CGO_ENABLED" = "1" ] && [ "$(go env GOOS)" != "windows" ] && [ "$DOCKER_STATIC" != "1" ]; then
+	# pkcs11 cannot be compiled statically if CGO is enabled (and glibc is used)
+	DOCKER_BUILDTAGS+=" pkcs11"
+fi
```

Comment:
This may have been from when the CLI was still in this repo.

But looks like still needed for containerd; see https://github.com/moby/moby/pull/44513#discussion_r1034626488

```
#22 [containerd-build 3/3] RUN --mount=from=containerd-src,src=/usr/src/containerd,rw     --mount=type=cache,target=/root/.cache <<EOT (set -e...)
#22 0.862 + bin/ctr
#22 1.921 + bin/containerd
#22 3.631 # github.com/containerd/containerd/cmd/containerd
#22 3.631 /usr/bin/ld: /tmp/go-link-508382094/000019.o: in function `New':
#22 3.631 /go/src/github.com/containerd/containerd/vendor/github.com/miekg/pkcs11/pkcs11.go:77: warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
```

Reply:
Huh??

> This may have been from when the CLI was still in this repo.

The string `pkcs11` does not appear _anywhere_ in the removed lines on this PR diff. _This is brand new code._ There is no `//go:build` or `//+build` constraint which references the `pkcs11` tag anywhere in the repo, vendor tree included.

> But looks like still needed for containerd

_make.sh isn't in the loop for building containerd_, nor is `$DOCKER_BUILDTAGS` respected by the containerd build in `Dockerfile`. And I couldn't find a build constraint which references the `pkcs11` tag anywhere in github.com/containerd/containerd, either.

Reply:
> The string pkcs11 does not appear anywhere in the removed lines on this PR diff. This is brand new code.

Ah! hm, yes, in that case "I don't know" (I saw your comment, and expect it was code copied/moved from one of the `hack/` scripts).


> make.sh isn't in the loop for building containerd, nor is $DOCKER_BUILDTAGS respected by the containerd build in Dockerfile. And I couldn't find a build constraint which references the pkcs11 tag anywhere in github.com/containerd/containerd, either.

Yeah, we won't need it in `hack/make.sh` for that, but we should check the containerd builds; the fact it isn't in their code repository could be an issue (their static build targets were broken various times, and still may be, e.g. https://github.com/containerd/containerd/issues/5824)

Reply:
No files in either github.com/moby/moby or github.com/containerd/containerd contain lines matching `^//go:build .*pkcs11` or `^//+build .*pkcs11`. What else is there to check?

Furthermore, the premise that the `hack/make.sh` code adding the `pkcs11` build tag is a well-intentioned attempt to fix static containerd builds makes no sense: the `pkcs11` tag is _added_ on _dynamically-linked_ builds.

Reply:
Mystery solved: [the build tag is referenced in the github.com/docker/cli build scripts.](https://github.com/docker/cli/blob/850fb6921cf7a208fd47d83074f27d0eb7ae268e/scripts/build/binary#L10-L12)